### ISSUE TITLE
translate Characters: improve perfomance and allow flexible number of characters

### ISSUE THREAD

**Issue by MoisesHer** [rapidsai/cudf#3543]
**Describe the bug**
The method NVStrings:translate(std::pair< unsigned, unsigned > *table, unsigned int count) does not allow to translate one character into several ones (decompose). For instance, "í"   into   "`i"
, or "萬" into " 萬 "
We also observed the performance of this task is slower (between 7x - 50x) compared to a custom CUDA tokenizer:
[https://docs.google.com/spreadsheets/d/1C4b_oM9lye_XFcqBELnyn2iogHbiLBjaBLpKohrQ__c/edit?usp=sharing](https://docs.google.com/spreadsheets/d/1C4b_oM9lye_XFcqBELnyn2iogHbiLBjaBLpKohrQ__c/edit?usp=sharing)

**Steps/Code to reproduce bug**
We created a repository running experiments for comparing cuStrings vs. our custom tokenizer:
[https://gitlab-master.nvidia.com/moisesh/bert_clean_text](https://gitlab-master.nvidia.com/moisesh/bert_clean_text)

**Expected behavior**
We expect to have the possibility of performing flexible character transformations, including the conversion of one character into 1,2,3 or more characters. For instance, sometimes we need to add spaces around some specific characters, such as Chinese characters.  

**Environment overview (please complete the following information)**
 - Environment: Ubuntu 18.08, Titan V

**Environment details**
<details><summary>Click here to see environment details</summary><pre>
     
     **git***
     commit 5a73e75be813e76f750f3ad1416d8cdc54906bbd (HEAD -> branch-0.11, origin/branch-0.11, origin/HEAD)
     Merge: 53478c10e cc7c9af8c
     Author: Mark Harris <mharris@nvidia.com>
     Date:   Wed Dec 4 10:05:53 2019 +1100
     
     Merge pull request #3500 from seunghwak/fea_ext_fill_repeat_string
     
     [REVIEW] Fill & repeat (incl tests) for strings
     **git submodules***
     b165e1fb11eeea64ccf95053e40f2424312599cc thirdparty/cub (v1.7.1)
     63f644be44201467e3938d59ed9d89cc8725c35d thirdparty/jitify (remotes/origin/feature/api_v2_v0.10)
     39125e0e476b960c2001f1ec76a3441335ff91b2 thirdparty/libcudacxx (0.8.1-94-g39125e0)
     08bc464bd8f4d779e4294305aa7dadebcebcc507 thirdparty/libcudacxx/libcxx (heads/rapidsai-interop)
     
     ***OS Information***
     DISTRIB_ID=Ubuntu
     DISTRIB_RELEASE=18.04
     DISTRIB_CODENAME=bionic
     DISTRIB_DESCRIPTION="Ubuntu 18.04.2 LTS"
     NAME="Ubuntu"
     VERSION="18.04.2 LTS (Bionic Beaver)"
     ID=ubuntu
     ID_LIKE=debian
     PRETTY_NAME="Ubuntu 18.04.2 LTS"
     VERSION_ID="18.04"
     HOME_URL="https://www.ubuntu.com/"
     SUPPORT_URL="https://help.ubuntu.com/"
     BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
     PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
     VERSION_CODENAME=bionic
     UBUNTU_CODENAME=bionic
     Linux moisesh-X299-A 5.0.0-36-generic #39~18.04.1-Ubuntu SMP Tue Nov 12 11:09:50 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
     
     ***GPU Information***
     Thu Dec  5 15:47:44 2019
     +-----------------------------------------------------------------------------+
     | NVIDIA-SMI 440.27       Driver Version: 440.27       CUDA Version: 10.2     |
     |-------------------------------+----------------------+----------------------+
     | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
     | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
     |===============================+======================+======================|
     |   0  Tesla T4            Off  | 00000000:04:00.0 Off |                    0 |
     | N/A   49C    P0    29W /  75W |   1007MiB / 15109MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   1  TITAN V             Off  | 00000000:17:00.0 Off |                  N/A |
     | 37%   53C    P2    37W / 250W |   1620MiB / 12066MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   2  GeForce GT 710      Off  | 00000000:65:00.0 N/A |                  N/A |
     | 50%   51C    P0    N/A /  N/A |    676MiB /  1993MiB |     N/A      Default |
     +-------------------------------+----------------------+----------------------+
     
     +-----------------------------------------------------------------------------+
     | Processes:                                                       GPU Memory |
     |  GPU       PID   Type   Process name                             Usage      |
     |=============================================================================|
     |    0     22372      C   trtserver                                    996MiB |
     |    1      6944      C   /usr/lib/libreoffice/program/soffice.bin     309MiB |
     |    1     22372      C   trtserver                                   1298MiB |
     |    2                    Not Supported                                       |
     +-----------------------------------------------------------------------------+
     
     ***CPU***
     Architecture:        x86_64
     CPU op-mode(s):      32-bit, 64-bit
     Byte Order:          Little Endian
     CPU(s):              12
     On-line CPU(s) list: 0-11
     Thread(s) per core:  2
     Core(s) per socket:  6
     Socket(s):           1
     NUMA node(s):        1
     Vendor ID:           GenuineIntel
     CPU family:          6
     Model:               85
     Model name:          Intel(R) Core(TM) i7-7800X CPU @ 3.50GHz
     Stepping:            4
     CPU MHz:             1200.081
     CPU max MHz:         4000.0000
     CPU min MHz:         1200.0000
     BogoMIPS:            7000.00
     Virtualization:      VT-x
     L1d cache:           32K
     L1i cache:           32K
     L2 cache:            1024K
     L3 cache:            8448K
     NUMA node0 CPU(s):   0-11
     Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req md_clear flush_l1d
     
     ***CMake***
     /usr/local/bin/cmake
     cmake version 3.15.0-rc3
     
     CMake suite maintained and supported by Kitware (kitware.com/cmake).
     
     ***g++***
     /usr/bin/g++
     g++ (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
     Copyright (C) 2017 Free Software Foundation, Inc.
     This is free software; see the source for copying conditions.  There is NO
     warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
     
     
     ***nvcc***
     
     ***Python***
     /usr/bin/python
     Python 2.7.15
     
     ***Environment Variables***
     PATH                            : /home/moisesh/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
     LD_LIBRARY_PATH                 :
     NUMBAPRO_NVVM                   :
     NUMBAPRO_LIBDEVICE              :
     CONDA_PREFIX                    :
     PYTHON_PATH                     :
     
     conda not found
     ***pip packages***
     /usr/bin/pip
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
     absl-py (0.8.1)
     appdirs (1.4.3)
     asn1crypto (0.24.0)
     astor (0.7.1)
     atomicwrites (1.3.0)
     attrs (19.1.0)
     autograd (1.2)
     aws-shell (0.2.1)
     awscli (1.16.215)
     backports-abc (0.5)
     backports.shutil-get-terminal-size (1.0.0)
     backports.weakref (1.0.post1)
     beautifulsoup4 (4.6.0)
     bleach (3.1.0)
     boto3 (1.9.205)
     botocore (1.12.205)
     certifi (2019.6.16)
     cffi (1.12.3)
     chardet (3.0.4)
     Click (7.0)
     cloudpickle (1.2.1)
     colorama (0.3.9)
     configobj (5.0.6)
     configparser (3.7.4)
     contextlib2 (0.5.5)
     cryptography (2.1.4)
     decorator (4.4.0)
     defusedxml (0.6.0)
     Django (1.11.26)
     docutils (0.14)
     entrypoints (0.3)
     enum34 (1.1.6)
     funcsigs (1.0.2)
     functools32 (3.2.3.post2)
     future (0.17.1)
     futures (3.3.0)
     gast (0.2.2)
     gluonnlp (0.7.0.dev0, /home/moisesh/Projects/Amazon/BERT-branch/mxnet/gluon-nlp/src)
     graphviz (0.11.1)
     grpcio (1.19.0)
     h5py (2.9.0)
     html5lib (0.9999999)
     idna (2.8)
     image (1.5.27)
     importlib-metadata (0.18)
     ipaddress (1.0.22)
     ipykernel (4.10.0)
     ipython (5.8.0)
     ipython-genutils (0.2.0)
     ipywidgets (7.4.2)
     Jinja2 (2.10.1)
     jmespath (0.9.4)
     joblib (0.13.2)
     jsonschema (3.0.1)
     jupyter (1.0.0)
     jupyter-client (5.2.4)
     jupyter-console (5.2.0)
     jupyter-core (4.4.0)
     Keras-Applications (1.0.7)
     Keras-Preprocessing (1.0.9)
     keyring (10.6.0)
     keyrings.alt (3.0)
     lxml (4.2.1)
     Mako (1.0.13)
     Markdown (3.1.1)
     MarkupSafe (1.1.1)
     mistune (0.8.4)
     mock (3.0.5)
     more-itertools (5.0.0)
     mxnet (1.6.0, /home/moisesh/Projects/Amazon/MKLDNN/incubator-mxnet/python)
     mxnet-cu101 (1.5.0)
     nbconvert (5.5.0)
     nbformat (4.4.0)
     nose (1.3.7)
     notebook (5.7.8)
     numpy (1.16.5)
     packaging (19.0)
     pandocfilters (1.4.2)
     pathlib2 (2.3.3)
     pbr (5.1.3)
     pexpect (4.7.0)
     pickleshare (0.7.5)
     Pillow (6.2.1)
     pip (9.0.1)
     pluggy (0.12.0)
     prometheus-client (0.6.0)
     prompt-toolkit (1.0.16)
     protobuf (3.10.0)
     psutil (5.6.3)
     ptyprocess (0.6.0)
     py (1.8.0)
     pyasn1 (0.4.6)
     pycparser (2.19)
     pycrypto (2.6.1)
     pycuda (2019.1)
     Pygments (2.4.2)
     pygobject (3.26.1)
     pyparsing (2.4.0)
     pyrsistent (0.15.2)
     pytest (4.6.4)
     python-dateutil (2.8.0)
     pytools (2019.1.1)
     pytz (2019.3)
     pyxdg (0.25)
     PyYAML (5.1)
     pyzmq (18.0.1)
     qtconsole (4.4.4)
     requests (2.22.0)
     rsa (3.4.2)
     s3transfer (0.2.1)
     sacremoses (0.0.34)
     scandir (1.10.0)
     scour (0.36)
     SecretStorage (2.3.1)
     Send2Trash (1.5.0)
     setuptools (41.4.0)
     simplegeneric (0.8.1)
     singledispatch (3.4.0.3)
     six (1.12.0)
     tensorboard (1.13.1)
     tensorflow-estimator (1.13.0)
     tensorflow-gpu (1.13.1)
     tensorflow-tensorboard (1.5.1)
     tensorrt (5.1.5.0)
     termcolor (1.1.0)
     terminado (0.8.2)
     testpath (0.4.2)
     tornado (5.1.1)
     tqdm (4.36.1)
     traitlets (4.3.2)
     urllib3 (1.25.3)
     wcwidth (0.1.7)
     webencodings (0.5.1)
     Werkzeug (0.16.0)
     wheel (0.33.6)
     widgetsnbextension (3.4.2)
     zipp (0.5.2)
     
</pre></details>

**Comment by kkraus14** [comment]
@davidwendt is this still relevant or should we close this?

**Comment by davidwendt** [comment]
We can probably close this. The original use case is now solved by the `nvtext::normalize_characters` API.
